### DISCUSSION TITLE
Guidance needed - One Job per Trigger or One Job for All Triggers

### DISCUSSION THREAD

**Discussion by ghost** [quartznet/quartznet#1678] [Q&A] [answered]
Hi,

I'm just looking for what the "Best Practice" is for my situation. I'm also trying to avoid any "gotchas" or technical issues that I am not aware of, having not yet implemented Quartz...

Currently we only have one class that implements IJob. But that one job will be run at many different times (many triggers).

What do I need to consider or watch out for, when deciding between these two options:

1. Create one Job and assign many different Triggers to that Job (100+).
2. Create a Job for each of Trigger?

NOTE: I will be assigning distinct values to each Trigger's JobDataMap, if that factors into the decision.

It seems unnecessary and wasteful to create a Job for each Trigger, since each job database row will be the same. Perhaps one consideration is that we may want to give some Jobs different names and organize them into different Job Groups, for some reason.

By the way, the Job will just run a .net console app, the console app returns an exitcode and when the console app finishes I set the JobExecutionContext.Result to the exitcode so it can be logged with the plugin. The console app may run for over 30 minutes at times...

**Comment by lahma** [comment, marked as answer]
> Create one Job and assign many different Triggers to that Job (100+).

As long as your job has one, well defined, responsibility, it should be OK. Like "Send me daily summary of X" - there can be as many triggers as there are users that have configured a custom time (just an example) without the fear that changing job to do other things would affect all users after the change. Basically, job implementation should be well defined.

> Create a Job for each of Trigger?

If you want different concurrency semantics maybe. Remember that each job key is how `DisallowConcurrentExecution` is handled. Register same implementation multiple times then the same job type might be running simultaneously, under different job key.

I think if you have well defined and constrained job implementation, one should suffice. Then you can categorize them with job groups to things that might need handling as group (like pausing) - groups can be like maintenance, internal, newsletter - whatever fits your needs. Trigger's job data map will allow the parametrization, but would it happen that for some reason job needs to be split into two, that would mean that you need to update affected triggers to refer to new job type/key.

**Reply by ghost** [reply]
Thanks for the insight and quick reply! Also, thanks for mentioned concurrency and pausing as things to consider At this point I don't plan on marking the job class with DisallowConcurrentExecution, as we will have no restrictions in that regard.
